### PR TITLE
Define a json schema for MML CREATE GENERATOR statements

### PR DESCRIPTION
**Pull request by asilversempirical**
...and build up some utility methods around it.

This implements everything but validation from https://docs.google.com/document/d/11iZf-yeCkXvkAXdIl5QPariLbbOJDRos_iB9WOMtWts/edit

### CODE REVIEW

File: src/mml_utils.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+from enum import Enum
+from string import Template
+
+import json
+import jsonschema
+import pkgutil
+
+# DO_NOT_SUBMIT: I'm pretty sure this all needs to be wrapped in a class
+# and read in the constructor, but make sure I'm not being paranoid.
+schema_json = pkgutil.get_data(
```
Comment (madeleine-empirical) [inline src/mml_utils.py]: I would not personally feel guilty about this.

File: src/mml.schema.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "title": "Schema for MML commands to CREATE GENERATOR",
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "type": "object",
+  "additionalProperties": false,
+  "required": ["metamodel", "columns"],
+  "properties": {
+    "metamodel": {"type": "string"},
```
Comment (madeleine-empirical) [inline src/mml.schema.json:8]: Am I understanding correctly that this the semantics of this is "the metamodel this was tested with"?

File: src/mml_utils.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+def to_mml(mml_json, table, generator):
+  jsonschema.validate(mml_json, MML_SCHEMA)
+  schema_phrase = ','.join(["%s %s" % (bql_quote_name(col), v['stattype']) for col,v
+                           in mml_json['columns'].items() if v['stattype'] is not 'IGNORE'])
+  subsample = mml_json.get('subsample', None)
+  return (Template('CREATE GENERATOR $gen FOR $table '
+                   'USING $metamodel($subsample $schema_phrase);')
+              .substitute(
+                  gen = generator,
+                  table = table,
```
Comment (madeleine-empirical) [inline src/mml_utils.py]: Needs more `bql_quote_name`.

File: tests/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import json
+import jsonschema
+import pkgutil
+
+def test_example_schema():
+    schema_json = pkgutil.get_data('bdbcontrib', 'mml.schema.json')
+    schema = json.loads(schema_json)
+    example = json.load(open('tests/example_mml_schema.json'))
+    jsonschema.validate(example, schema)
```
Comment (madeleine-empirical) [inline tests/test_schema.py]: Check example for correctness?

Comment (madeleine-empirical) [inline tests/test_schema.py]: Oops, I'm confused.

File: tests/test_mml_utils.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        guesses = mml_utils.guess_types(bdb, 't')
+        assert guesses == (
+                {'col1': StatType.IGNORE,
+                'col2': StatType.CATEGORICAL,
+                'col3': StatType.IGNORE,
+                'col4': StatType.NUMERICAL,
+                'col5': StatType.CATEGORICAL})
+
+        mml_json = mml_utils.to_json(guesses)
+        assert mml_json == {'metamodel': 'crosscat',
```
Comment (madeleine-empirical) [inline tests/test_mml_utils.py]: This is some funky indentation here.

File: tests/test_mml_utils.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+# Because this test may indirectly import pylab, which would peg the
+# matplotlib backend, which might prevent a later test from drawing
+# pictures headless.  &#&*%(@#^&!@.
+import matplotlib
+matplotlib.use('Agg')
```
Comment (madeleine-empirical) [inline tests/test_mml_utils.py]: What's matplotlib doing here?

**Comment by madeleine-empirical** [comment]
Aside from the incompleteness, looks good.

File: src/mml.schema.json (hunk 2 above)
Comment (asilversempirical) [inline src/mml.schema.json:8]: No. The intent is that this schema is for all arguments to CREATE GENERATOR after USING. So everything besides table and generator name. It's admittedly a weird split, and may change. I'm not totally sure whether the validator will take a schema or just the guesses.

File: tests/test_mml_utils.py (hunk 6 above)
Comment (asilversempirical) [inline tests/test_mml_utils.py]: I was taking that comment from the other files at its word, but removing it doesn't seem to break anything. Removed.

File: tests/test_mml_utils.py (hunk 5 above)
Comment (asilversempirical) [inline tests/test_mml_utils.py]: I immediately lose my mind without pep8 automatically checking.

**Comment by asilversempirical** [comment]
I think the main incompleteness at this point is the lack of validate. But I think this is useful without even without that, and would like to split this and validate into two PRs to probcomp. Is there another source of incompleteness you mean?

File: src/mml.schema.json (hunk 2 above)
Comment (madeleine-empirical) [inline src/mml.schema.json:8]: So, even without the testing, it's "I mean these stattypes relative to that generator." So you wouldn't put "poisson" in a Lovcat spec, but you would put it in a gpmcc spec?

File: src/mml.schema.json [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    }
+  },
+  "definitions": {
+    "column_def": {
+      "type": "object",
+      "required": ["stattype"],
+      "properties": {
+          "stattype": { "$ref": "#/definitions/stattype" },
+          "guess": { "$ref": "#/definitions/stattype" }
+          }
```
Comment (madeleine-empirical) [inline src/mml.schema.json]: funky indentation

File: src/mml_utils.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import pkgutil
+
+MML_SCHEMA = json.loads(
+    pkgutil.get_data('bdbcontrib', 'mml.schema.json'))
+
+
+class StatType(Enum):
+    CATEGORICAL = 1
+    NUMERICAL = 2
+    IGNORE = 3
```
Comment (madeleine-empirical) [inline src/mml_utils.py:17]: TIL

**Comment by madeleine-empirical** [comment]
Send this as a PR to probcomp at your convenience.

**Comment by asilversempirical** [comment]
Opened https://github.com/probcomp/bdbcontrib/pull/153. Closing this one out

**Pull request closed without merging**